For #7836: Refine add-ons badge design

hakkikaancaliskan opened the pull request:
cc @brampitoyo

https://github.com/mozilla-mobile/fenix/issues/12845#issuecomment-663272360

Colors are coming from extensions. They need to set more stylish colors.

<img src="https://user-images.githubusercontent.com/17825767/88461470-61e02e00-ceac-11ea-99a4-822e3bc05d8f.png" width="200"><img src="https://user-images.githubusercontent.com/17825767/88461472-63115b00-ceac-11ea-968e-3cee2eea8d7f.png" width="200"><img src="https://user-images.githubusercontent.com/17825767/88461474-63115b00-ceac-11ea-90a8-ba043509c1e6.png" width="200">
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

cadeyrn commented:
Thanks for working on it. Apart from the colours I noticed some other things that are different from the mockup: 

- The width is different (width should be equal to height, in your screenshot it has a bigger width than height).
- The border radius is different (your screenshot seems to have a larger border radius than the mockup).
- The font weight is different (it should be bold, it's not bold in your screenshot).
- The font size is different (not a lot, but it's a bit bigger in your screenshot compared to the mockup).

hakkikaancaliskan commented:
> Thanks for working on it. Apart from the colours I noticed some other things that are different from the mockup: 
> 
> - The width is different (width should be equal to height, in your screenshot it has a bigger width than height).
> - The border radius is different (your screenshot seems to have a larger border radius than the mockup).
> - The font weight is different (it should be bold, it's not bold in your screenshot).
> - The font size is different (not a lot, but it's a bit bigger in your screenshot compared to the mockup).

Thanks for feedback. That's why this PR is draft :)
Actually I'm waiting for a ux mock to how exactly it should be @brampitoyo

hakkikaancaliskan commented:
When i run `./gradlew test` i get unrelated test fails and can't get menu fail.

[index.zip](https://github.com/mozilla-mobile/android-components/files/4976042/index.zip)

Amejia481 commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/item/WebExtensionBrowserMenuItem.kt:
```diff
@@ -41,10 +41,10 @@ class WebExtensionBrowserMenuItem(
 
     @Suppress("TooGenericExceptionCaught")
     override fun bind(menu: BrowserMenu, view: View) {
-        val imageView = view.findViewById<ImageView>(R.id.action_image)
-        val labelView = view.findViewById<TextView>(R.id.action_label)
-        val badgeView = view.findViewById<TextView>(R.id.badge_text)
-        val container = view.findViewById<View>(R.id.container)
```
Nit: Is there any specific reason for the change? Performance or other reason? :)

hakkikaancaliskan replied:
Nothing lol
Just looks nice to me :)

Amejia481 replied:
😄  I think if we don't have a strong reason for the change, I would prefer to not change it? What do you think? :)

hakkikaancaliskan replied:
Maybe our reason could be "this code is more like 'kotlin-style'" lol
But if you want i can take back changes.

Amejia481 commented:
In my case, I prefer only to run tests for the module that I'm modifying in the moment and open a pull request to let the CI run all the rest. For running test for `browser-menu` will be only `./gradlew -Pcoverage browser-menu:test`

It looks the only tests that are failing are related to `browser-menu` it could linked to the `background.setTint(it)` change.

<img width="1052" alt="image" src="https://user-images.githubusercontent.com/773158/88579812-a5f04180-d018-11ea-8ab5-85f124e90504.png">

![image](https://user-images.githubusercontent.com/773158/88578903-42b1df80-d017-11ea-9667-7410ab1560a0.png)

You can see the logs for an module just by clicking the module name `build-browser-menu`  -> `Details` -> `View task in Taskcluster` -> `View Live Log`

[browser-menu logs](https://firefoxci.taskcluster-artifacts.net/A3Kon1tWTbuRDCoP0PKfLA/0/public/logs/live_backing.log )

hakkikaancaliskan commented:
@Amejia481 sorry but I'm unable to figure out what is wrong with test. I'm not good with tests actually. Why does `action.badgeBackgroundColor?.let { badgeView.background.setTint(it) }` throwing null pointer?
```
java.lang.NullPointerException
	at mozilla.components.browser.menu.item.WebExtensionBrowserMenuItem.bind(WebExtensionBrowserMenuItem.kt:57)
	at mozilla.components.browser.menu.item.WebExtensionBrowserMenuItemTest.badge text view is invisible if action badge text is empty(WebExtensionBrowserMenuItemTest.kt:157)
```

Amejia481 commented:
> @Amejia481 sorry but I'm unable to figure out what is wrong with test. I'm not good with tests actually. Why does `action.badgeBackgroundColor?.let { badgeView.background.setTint(it) }` throwing null pointer?
> 
> ```
> java.lang.NullPointerException
> 	at mozilla.components.browser.menu.item.WebExtensionBrowserMenuItem.bind(WebExtensionBrowserMenuItem.kt:57)
> 	at mozilla.components.browser.menu.item.WebExtensionBrowserMenuItemTest.badge text view is invisible if action badge text is empty(WebExtensionBrowserMenuItemTest.kt:157)
> ```

@hakkikaancaliskan   I think it's failing because we are now calling `badgeView.background` and `background` could be null that it's the case of the tests. If we call `badgeView.background?.setTint(it)` instead, it will be safer as we are not assuming that we always have a `background` :)

Amejia481 commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/item/WebExtensionBrowserMenuItem.kt:
```diff
@@ -54,7 +54,7 @@ class WebExtensionBrowserMenuItem(
         badgeView.setBadgeText(action.badgeText)
         action.badgeTextColor?.let { badgeView.setTextColor(it) }
-        action.badgeBackgroundColor?.let { badgeView.setBackgroundColor(it) }
+        action.badgeBackgroundColor?.let { badgeView.background.setTint(it) }
```
Let's have a test to verify that every time we have a `badgeBackgroundColor` we are setting the tint.

hakkikaancaliskan replied:
I've tried to write that test but badgeView.background returns null and i can't test it. What should i do?

![image](https://user-images.githubusercontent.com/17825767/88718358-89dcc500-d12a-11ea-8523-b97225e53c65.png)

Amejia481 replied:
It's `null` because in the test `badgeView` is mock and the background is not provided, you can create a real object         `val badgeView = TextView(testContext)` in the test

hakkikaancaliskan replied:
I've changed val badgeView: TextView = spy(TextView(testContext)) with val badgeView = TextView(testContext) but no luck. It's null.

Amejia481 replied:
I think you have to assign a value as by default is null :)

hakkikaancaliskan replied:
I've assigned now but Line 195 throwing null :(
Could you please take a look, this is driving me crazy

Amejia481 commented on components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserMenuItemTest.kt:
```diff
@@ -161,6 +162,40 @@ class WebExtensionBrowserMenuItemTest {
+        testDispatcher.advanceUntilIdle()
+
+        badgeView.background?.setTint(Color.BLUE)
+        val badgeViewBackground = badgeView.background as ColorDrawable
```
Oh, here the background is not getting initialized only casted.

Amejia481 replied:
Do not worry I noticed that the bind test was already testing this functionality

Amejia481 replied:
Take a look at to the snipped above the same solution could be applied here.

hakkikaancaliskan replied:
Ohhh, we're gonna remove this test? Ok, how we're gonna improve bind test?

Amejia481 replied:
The test looks good we can keep it, it would be an additional check. But we definitely need to improve the test `bind` as now it not verifying see `verify(badgeView).background?.setTint(Color.BLUE)` the `?`  is hiding the check

Amejia481 commented on components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserMenuItemTest.kt:
```diff
@@ -119,13 +120,13 @@ class WebExtensionBrowserMenuItemTest {
+        verify(badgeView).text = "badgeText"
         verify(badgeView).setTextColor(Color.WHITE)
-        verify(badgeView).setBackgroundColor(Color.BLUE)
+        verify(badgeView).background?.setTint(Color.BLUE)
```
This one is already testing the use case when we pass a background it needs a bit of work

Amejia481 replied:
You could take a look this snipped, here the background is getting initialized before it being used :)
Hope it helps!
 
```kotlin
   @Test
    fun bind() {
        val icon: Bitmap = mock()
        val imageView: ImageView = mock()
        val badgeView = spy(TextView(testContext))
        val background : Drawable = mock()
        val labelView: TextView = mock()
        val container = View(testContext)
        val view: View = mock()

        whenever(badgeView.background).thenReturn(background)
        whenever(view.findViewById<ImageView>(R.id.action_image)).thenReturn(imageView)
        whenever(view.findViewById<ImageView>(R.id.action_image)).thenReturn(imageView)
        whenever(view.findViewById<TextView>(R.id.badge_text)).thenReturn(badgeView)
        whenever(view.findViewById<TextView>(R.id.action_label)).thenReturn(labelView)
        whenever(view.findViewById<View>(R.id.container)).thenReturn(container)
        whenever(view.context).thenReturn(testContext)

        val browserAction = Action(
            title = "title",
            loadIcon = { icon },
            enabled = true,
            badgeText = "badgeText",
            badgeTextColor = Color.WHITE,
            badgeBackgroundColor = Color.BLUE
        ) {}

        val action = WebExtensionBrowserMenuItem(browserAction) {}
        action.bind(mock(), view)
        testDispatcher.advanceUntilIdle()

        val iconCaptor = argumentCaptor<BitmapDrawable>()
        verify(imageView).setImageDrawable(iconCaptor.capture())
        assertEquals(icon, iconCaptor.value.bitmap)

        verify(imageView).contentDescription = "title"
        verify(labelView).text = "title"
        verify(badgeView).text = "badgeText"
        verify(badgeView).setTextColor(Color.WHITE)
        verify(background).setTint(Color.BLUE)
    }
```

hakkikaancaliskan replied:
So i'm gonna delete test that i wrote?

hakkikaancaliskan replied:
Ok, done 🎉
Could you please verify now that test working properly?

Amejia481 commented on components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserMenuItemTest.kt:
```diff
@@ -119,13 +121,13 @@ class WebExtensionBrowserMenuItemTest {
+        verify(badgeView).text = "badgeText"
         verify(badgeView).setTextColor(Color.WHITE)
-        verify(badgeView).setBackgroundColor(Color.BLUE)
+        verify(background).setTint(Color.BLUE)
```
👍

hakkikaancaliskan commented:
I think it's done. Could you please check?

Amejia481 commented:
The code looks good to me :) Let's see what the CI says XD 
Also we need the feedback from @brampitoyo before landing, that final touch will be to squash all the commits into a single one

hakkikaancaliskan commented:
> 
> 
> The code looks good to me :) Let's see what the CI says XD
> Also we need the feedback from @brampitoyo before landing, that final touch will be to squash all the commits into a single one

👍
I hate tests xd

Amejia481 commented:
> > The code looks good to me :) Let's see what the CI says XD
> > Also we need the feedback from @brampitoyo before landing, that final touch will be to squash all the commits into a single one
> 
> 👍
> I hate tests xd

Nobody love them but after a while you get used to have a good relationship 😉

hakkikaancaliskan commented:
@Amejia481 CI passed. So am i going to squash commits? Because maybe i cant find time for squashing them :)

brampitoyo commented:
@hakkikaancaliskan I’ve posted the badge specs on https://github.com/mozilla-mobile/fenix/issues/12845#issue-664198860

hakkikaancaliskan commented:
Current look (20dp min width, 10sp font size):
![Screenshot_20200729-181501_Sample_Browser](https://user-images.githubusercontent.com/17825767/88818912-f4900e00-d1c7-11ea-92b3-469a98884cfe.png)

hakkikaancaliskan commented:
@Amejia481 https://github.com/mozilla-mobile/fenix/issues/12845#issuecomment-667794419
UX approved :)

hakkikaancaliskan commented:
@Amejia481 bip bip

Amejia481 commented:
Sorry for the delay, great job!

hakkikaancaliskan commented:
> Sorry for the delay, great job!

No problem :) squash commits needed?

hakkikaancaliskan commented:
> Sorry for the delay, great job!

Oops, you guys working on Sunday too?

Amejia481 commented:
bors r-

Amejia481 commented:
> > Sorry for the delay, great job!
> 
> No problem :) squash commits needed?

Yes, please :)

Amejia481 commented:
> > Sorry for the delay, great job!
> 
> Oops, you guys working on Sunday too?

Not but I felt guilty as didn't review before 😅, do not worries it's not a problem.

hakkikaancaliskan commented:
> > > Sorry for the delay, great job!
> > 
> > 
> > Oops, you guys working on Sunday too?
> 
> Not but I felt guilty as didn't review before 😅, do not worries it's not a problem.

Ohh of course this could wait monday. Go bro go, enjoy weekend 😂

Amejia481 commented:
> Ohh of course this could wait monday. Go bro go, enjoy weekend 😂

Thanks :),  really don't worry after squashing the commit it will be just landing :)

hakkikaancaliskan commented:
> > Ohh of course this could wait monday. Go bro go, enjoy weekend 😂
> 
> Thanks :), really don't worry after squashing the commit it will be just landing :)

Squashed and rebased 😊

Amejia481 commented:
bors r+

Amejia481 commented:
> > > Ohh of course this could wait monday. Go bro go, enjoy weekend 😂
> > 
> > 
> > Thanks :), really don't worry after squashing the commit it will be just landing :)
> 
> Squashed and rebased 😊

Thanks :)
Hope you have a great rest of the weekend!

hakkikaancaliskan commented:
> > > > Ohh of course this could wait monday. Go bro go, enjoy weekend 😂
> > > 
> > > 
> > > Thanks :), really don't worry after squashing the commit it will be just landing :)
> > 
> > 
> > Squashed and rebased 😊
> 
> Thanks :)
> Hope you have a great rest of the weekend!

Hope you too :)
